### PR TITLE
MM-46494: Add filter options to browse channel modal

### PR DESCRIPTION
**Pull request by sinansonmez**
#### Summary
This PR is adding two new filtering options (all channels and private channels) to browse channel modal.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost/issues/21172
Jira https://mattermost.atlassian.net/browse/MM-46494

#### Screenshots
|  before  |  after  |
|----|----|
| Screenshot 2023-07-23 at 10 12 02 | Screenshot 2023-07-23 at 10 10 58 |

#### Release Note
```release-note
Added two new filtering options (show all channel types, show private channels to browse channel modal.
```

### CODE REVIEW

**Comment by M-ZubairAhmed** [comment]
@sinansonmez can you resolve the CI

**Comment by sinansonmez** [comment]
@M-ZubairAhmed resolved

File: webapp/channels/src/components/searchable_channel_list.tsx [hunk 1]
```diff
@@ -371,52 +403,74 @@ export default class SearchableChannelList extends React.PureComponent<Props, St
...
+                size={18}
+                color={'var(--button-bg)'}
+            />
+        );
+        const channelDropdownItems = [
+            <Menu.ItemAction
+                key='channelsMoreDropdownAll'
+                id='channelsMoreDropdownAll'
+                onClick={() => this.props.changeFilter(FILTER.all)}
+                icon={<GlobeIcon size={16}/>} //  todo use the correct icon from compass: PR opened: https://github.com/mattermost/compass-icons/pull/78/files
```
Comment (sinansonmez) [inline webapp/channels/src/components/searchable_channel_list.tsx:417]: Attention

**Review by pvev — APPROVED** [review]
Thank you @sinansonmez , in general code looks good 👍 , just some small comments on it. Great work 🚀

**Comment by esethna** [comment]
@sinansonmez I think the kira ticket for this is: https://mattermost.atlassian.net/browse/MM-46494

Please let us know what is still left to do in a follow-up PR re sorting so QA knows what pieces to test. I'd propose we break the ticket up and file a new one for the whatever sorting isn't implemented. 

@johndavidlugtu I'll let you take this since in your teams AORs

**Comment by sinansonmez** [comment]
@yasserfaraazkhan I see that some e2e tests are failing. I fixed them but still there might be some failing. Could you please re-run test and I can check again?

File: webapp/channels/src/components/searchable_channel_list.tsx [hunk 2]
```diff
@@ -264,33 +259,70 @@ export default class SearchableChannelList extends React.PureComponent<Props, St
...
-                    }}
+                    values={{text: this.state.channelSearchValue}}
+                />
+            );
+        }
+        switch (this.props.filter) {
+        case 'archived':
+            return (
+                <FormattedMessage
+                    id={t('more_channels.noArchived')}
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/searchable_channel_list.tsx:277]: This t function need not be there here

File: webapp/channels/src/components/browse_channels/browse_channels.tsx [hunk 3]
```diff
@@ -212,13 +219,27 @@ export default class BrowseChannels extends React.PureComponent<Props, State> {
...
+        }
+        if (this.state.filter === FILTER.public) {
+            searchedChannels = channels.filter((c) => c.type === Constants.OPEN_CHANNEL && c.delete_at === 0);
+        }
+        if (this.state.filter === FILTER.archived) {
+            searchedChannels = channels.filter((c) => c.delete_at !== 0);
+        }
+        if (this.props.shouldHideJoinedChannels) {
+            searchedChannels = this.getChannelsWithoutJoined(searchedChannels);
+        }
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/browse_channels/browse_channels.tsx:235]: nit : can this be if else ladder the last else being https://github.com/mattermost/mattermost/pull/24099/files#diff-b0b14d135dc7be9da8fb91320dd550821ed8ef4c29393a1bfbd4ccdb9cfa3dd2R223

File: webapp/channels/src/components/searchable_channel_list.tsx [hunk 4]
```diff
@@ -264,33 +259,70 @@ export default class SearchableChannelList extends React.PureComponent<Props, St
...
+    getFilterLabel = () => {
+        switch (this.props.filter) {
+        case 'archived':
+            return <span>{localizeMessage('more_channels.show_archived_channels', 'Channel Type: Archived')}</span>;
+        case 'public':
+            return <span>{localizeMessage('more_channels.show_public_channels', 'Channel Type: Public')}</span>;
+        case 'private':
+            return <span>{localizeMessage('more_channels.show_private_channels', 'Channel Type: Private')}</span>;
+        default:
+            return <span>{localizeMessage('more_channels.show_all_channels', 'Channel Type: All')}</span>;
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/searchable_channel_list.tsx:317]: if you want the span you can just write <FormattedMessage/> , span are included in this.

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

**Comment by yasserfaraazkhan** [comment]
> Successfully triggered E2E testing!
> [GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/356081) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/356081)

thank you for fixing tests @sinansonmez, 
these 2 look like last remaining fixes needed for e2e.

MM-19337 Enable users to view archived channels
MM-T1467 Accessibility Support in More Channels Dialog screen

**Comment by johndavidlugtu** [comment]
Does this have any performance implications? Specifically wanted to know if there's any performance concerns on this functionality

**Comment by sinansonmez** [comment]
thanks @yasserfaraazkhan 

Thanks to test case`MM-19337 Enable users to view archived channels` I realized an unexpected behavior. 
In this specific test case, normal user (non-admin) is doing a search and receiving the following error from `https://8065-mattermost-mattermostgi-08nqm51cp9u.ws-eu102.gitpod.io/api/v4/channels/search?include_deleted=true` . 

If I add `nonAdminSearch: true` to api call [in this line](https://github.com/mattermost/mattermost/blob/01e576a6d451101d094ca26c1b7f38c5c67715bc/webapp/channels/src/components/browse_channels/browse_channels.tsx#L201), it works. However, this time, it return all channels from all teams which I dont need in this case. (of course I can filter the channels based on current team but it is additional computation on the client side) 

Therefore the question is that is there any way to use this endpoint so non-admin users can search only the channels from the current team without any client side filtering?  @pvev @M-ZubairAhmed any idea?

```
{
    "id": "api.context.permissions.app_error",
    "message": "You do not have the appropriate permissions.",
    "detailed_error": "userId=4bepga9qopynpmp5tm9xizsfrr, permission=sysconsole_read_user_management_channels,",
    "request_id": "5xcs6rm1gigyigxnxes85hokiw",
    "status_code": 403
}
```

File: webapp/channels/src/components/searchable_channel_list.tsx (hunk 1 above)
Comment (sinansonmez) [reply webapp/channels/src/components/searchable_channel_list.tsx:417]: @M-ZubairAhmed before doing that I think compass icons package needs to be published first. I dont think I can do it as community member.

File: webapp/channels/src/components/searchable_channel_list.tsx [hunk 5]
```diff
@@ -371,51 +403,71 @@ export default class SearchableChannelList extends React.PureComponent<Props, St
...
+                ariaLabel={localizeMessage('suggestion.all', 'All channel types')}
+            />,
+            <Menu.ItemAction
+                key='channelsMoreDropdownPublic'
+                id='channelsMoreDropdownPublic'
+                onClick={() => this.props.changeFilter(FILTER.public)}
+                icon={<GlobeIcon size={16}/>}
+                text={localizeMessage('suggestion.public', 'Public channels')}
+                rightDecorator={this.props.filter === FILTER.public ? checkIcon : null}
+                ariaLabel={localizeMessage('suggestion.public', 'Public channels')}
```
[the remark this answers is not in the document]
Comment (sinansonmez) [reply webapp/channels/src/components/searchable_channel_list.tsx:429]: Sure, but just to double check do you want me to replace every `localizeMessage` with `formatMessage` in `browse_channels` and `searchable_channel_list` files?

File: webapp/channels/src/components/searchable_channel_list.tsx [hunk 6]
```diff
@@ -371,51 +403,71 @@ export default class SearchableChannelList extends React.PureComponent<Props, St
...
+                    id='channelsMoreDropdownArchived'
+                    onClick={() => this.props.changeFilter(FILTER.archived)}
+                    icon={<ArchiveOutlineIcon size={16}/>}
+                    text={localizeMessage('suggestion.archive', 'Archived channels')}
+                    rightDecorator={this.props.filter === FILTER.archived ? checkIcon : null}
+                    ariaLabel={localizeMessage('suggestion.archive', 'Archived channels')}
+                />,
             );
         }
+        const channelDropdown = (
```
[the remark this answers is not in the document]
Comment (sinansonmez) [reply webapp/channels/src/components/searchable_channel_list.tsx:454]: Sure let me take a look at the new component.

File: webapp/channels/src/components/searchable_channel_list.tsx (hunk 4 above)
Comment (sinansonmez) [reply webapp/channels/src/components/searchable_channel_list.tsx:317]: Sorry @M-ZubairAhmed it is not clear to me where is span included?

File: webapp/channels/src/components/searchable_channel_list.tsx (hunk 2 above)
Comment (sinansonmez) [reply webapp/channels/src/components/searchable_channel_list.tsx:277]: thanks updated as suggested

File: webapp/channels/src/components/searchable_channel_list.tsx [hunk 7]
```diff
@@ -264,33 +259,70 @@ export default class SearchableChannelList extends React.PureComponent<Props, St
...
+        switch (this.props.filter) {
+        case 'archived':
+            return (
+                <FormattedMessage
+                    id={t('more_channels.noArchived')}
+                    tagName='strong'
+                    defaultMessage={'No archived channels'}
+                />
+            );
+        case 'private':
```
[the remark this answers is not in the document]
Comment (sinansonmez) [reply webapp/channels/src/components/searchable_channel_list.tsx:282]: fair point, updated as suggested

File: webapp/channels/src/components/browse_channels/browse_channels.tsx [hunk 8]
```diff
@@ -231,35 +252,37 @@ export default class BrowseChannels extends React.PureComponent<Props, State> {
...
-            this.activeChannels = search ? searchedChannels : archivedChannelsWithoutJoined;
-        } else if (shouldShowArchivedChannels && !shouldHideJoinedChannels) {
-            this.activeChannels = search ? searchedChannels : archivedChannels;
-        } else if (!shouldShowArchivedChannels && shouldHideJoinedChannels) {
-            this.activeChannels = search ? searchedChannels : otherChannelsWithoutJoined;
-        } else {
-            this.activeChannels = search ? searchedChannels : channels;
+    getChannelsWithoutJoined = (channelList: Channel[]) => channelList.filter((channel) => !this.isMemberOfChannel(channel.id));
+
+    getActiveChannels = () => {
```
[the remark this answers is not in the document]
Comment (sinansonmez) [reply webapp/channels/src/components/browse_channels/browse_channels.tsx:257]: great advice, much less code 😄

File: webapp/channels/src/components/browse_channels/browse_channels.tsx [hunk 9]
```diff
@@ -29,12 +29,19 @@ import './browse_channels.scss';
 const CHANNELS_CHUNK_SIZE = 50;
 const CHANNELS_PER_PAGE = 50;
 const SEARCH_TIMEOUT_MILLISECONDS = 100;
+export const FILTER = {
```
[the remark this answers is not in the document]
Comment (sinansonmez) [reply webapp/channels/src/components/browse_channels/browse_channels.tsx:32]: good point, updated as suggested

**Comment by sinansonmez** [comment]
> Does this have any performance implications? Specifically wanted to know if there's any performance concerns on this functionality

@johndavidlugtu I dont there will be performance issue as I am not introducing a new endpoint. I am simply calling different endpoint when searching the results. However, I ask internal team members to confirm @pvev @M-ZubairAhmed

File: webapp/channels/src/components/searchable_channel_list.tsx (hunk 1 above)
Comment (sinansonmez) [reply webapp/channels/src/components/searchable_channel_list.tsx]: @M-ZubairAhmed I see that Matt has published a new version if compass icon package. I upgraded the package and used the correct icon

**Comment by sinansonmez** [comment]
> thanks @yasserfaraazkhan
> 
> Thanks to test case`MM-19337 Enable users to view archived channels` I realized an unexpected behavior. In this specific test case, normal user (non-admin) is doing a search and receiving the following error from `https://8065-mattermost-mattermostgi-08nqm51cp9u.ws-eu102.gitpod.io/api/v4/channels/search?include_deleted=true` .
> 
> If I add `nonAdminSearch: true` to api call [in this line](https://github.com/mattermost/mattermost/blob/01e576a6d451101d094ca26c1b7f38c5c67715bc/webapp/channels/src/components/browse_channels/browse_channels.tsx#L201), it works. However, this time, it return all channels from all teams which I dont need in this case. (of course I can filter the channels based on current team but it is additional computation on the client side)
> 
> Therefore the question is that is there any way to use this endpoint so non-admin users can search only the channels from the current team without any client side filtering? @pvev @M-ZubairAhmed any idea?
> 
> ```
> {
>     "id": "api.context.permissions.app_error",
>     "message": "You do not have the appropriate permissions.",
>     "detailed_error": "userId=4bepga9qopynpmp5tm9xizsfrr, permission=sysconsole_read_user_management_channels,",
>     "request_id": "5xcs6rm1gigyigxnxes85hokiw",
>     "status_code": 403
> }
> ```

@M-ZubairAhmed @pvev for now I ended up filtering channels based on current team in the client side. Please let me know if I am missing something or is there any other endpoint that I can use. 

@yasserfaraazkhan with latest changes I should fix the last 2 failing e2e tests. Could you please re-run the tests?

File: webapp/channels/src/components/searchable_channel_list.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (sinansonmez) [reply webapp/channels/src/components/searchable_channel_list.tsx:482]: @M-ZubairAhmed old menu component is replaced with new menu component

**Comment by yasserfaraazkhan** [comment]
@sinansonmez `channel/archived_channels_1_spec.js` & `browse_channels_spec.js` file might need some more lint fixes. after that, we'll be able to run the e2e

**Comment by yasserfaraazkhan** [comment]
thank you @sinansonmez , the smoke test are fixed, can you check this file once `archived_channels_1_spec.js `.

**Comment by sinansonmez** [comment]
@yasserfaraazkhan thanks for the info. I hope this time, the last failing test is also fixed.

**Review by yasserfaraazkhan — APPROVED** [review]
thank you for the changes @sinansonmez ,

just one small look at `more_public_channels_spec.js ` seems it needs a small fix

**Comment by sinansonmez** [comment]
@yasserfaraazkhan thanks. FYI I also renamed the file name to align with the current naming of the modal (i.e Browse Channel Modal)

**Review by pvev — APPROVED** [review]
Thank you @sinansonmez , great work here! :shipit: 🚢 🚀

**Review by abhijit-singh — CHANGES_REQUESTED** [review]
Thank you @sinansonmez, looks great! 

There's one request — I don't think we should be including `Archived` channels when the `all channel types` filter is selected. If we do show them then with the current sorting, some archived channels might always appear on top, which is not ideal. This is also defined in the [design file](https://www.figma.com/file/KfzKLsROKCZ7zm4S5Sdi7X/MM-42267-%22Browse-Channels%22-Modal?type=design&node-id=239-115543&mode=design&t=e736UJll1k9WEOpP-4) linked in the Jira ticket (also attaching a screenshot).

image


The one other thing I'd like to add is a simple menu divider in the filter dropdown menu between the other options and `archived channels`. This can help in setting expectations about not showing archived channels when `All channel types` is selected. Adding a screenshot below - 
image

**Comment by sinansonmez** [comment]
Thanks for the feedback @abhijit-singh I will integrate the changes on the upcoming days

**Comment by sinansonmez** [comment]
@yasserfaraazkhan could you please re-run e2e tests?

**Comment by sinansonmez** [comment]
@abhijit-singh I implemented the changes. Please take a look

**Comment by yasserfaraazkhan** [comment]
@sinansonmez there I scenario, I want to share,

The member count In 'Browse channel modal' is not getting updated when user leaves the channel and opens Browse channel modal.   

example channel name : **cwqevxgmtdf**
This has 1 User. After we leave this and open browse channel , we can see the Count is till 1

Screenshot 2023-08-19 at 6 12 47 PM

Screenshot 2023-08-19 at 6 13 16 PM

When We Reload screen the current count get displayed.
Screenshot 2023-08-19 at 6 13 37 PM

**Comment by sinansonmez** [comment]
good catch @yasserfaraazkhan. This behavior should only happen when channel has only 1 member and that member leaves the channel. If you agree I would address that issue in a separate PR as the issue is not directly related to changes in this PR. Please open the issue and assign it to me. 

For this PR, I dont see any e2e failing test related to this PR, right?

**Comment by sinansonmez** [comment]
@M-ZubairAhmed @abhijit-singh please take a look at the latest changes

**Review by yasserfaraazkhan — APPROVED** [review]
Thank you @sinansonmez 
e2e have passed and functionality is working.
Just have 1 scenario as mentioned above.

**Review by abhijit-singh — APPROVED** [review]
Thank you @sinansonmez!

**Review by johndavidlugtu — APPROVED** [review]
Awesome, thanks @sinansonmez

File: webapp/channels/src/components/searchable_channel_list.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/searchable_channel_list.tsx]: No chaning for all occurence would be out of scope of this PR, please only change those places which this PR touched

**Review by M-ZubairAhmed — APPROVED** [review]
This is looking great Sinan, i just have the formatMessage nit pick that's all

File: webapp/channels/src/components/searchable_channel_list.tsx (hunk 4 above)
Comment (sinansonmez) [reply webapp/channels/src/components/searchable_channel_list.tsx]: Good point, updated as suggested

**Comment by hmhealey** [comment]
@M-ZubairAhmed Can this be merged? I'm not sure if it's waiting on anything else

**Pull request merged**